Further clean up after Nov 2019 network upgrade activated

sickpig opened the pull request:
Main content is the removal of NULLDUMMY from the code base. Since Schnorr multisig upgrade is active null dummy element has been re-purposed so there's no need for NULLDUMMY anymore. This part is a port of:

https://reviews.bitcoinabc.org/D4577
https://github.com/bitcoin/bitcoin/pull/10699/files (partial)

Still TODO:

https://reviews.bitcoinabc.org/D4673 (add SCHNORR_MULTISIG to mandatory flags)
https://reviews.bitcoinabc.org/D4679 (add SCRIPT_VERIFY_MINIMALDATA to mandatory flags)
https://reviews.bitcoinabc.org/D4576 (drop 'check3' upgrade-conditional-script-failure for Schnorr multisig) 

Rebased on top of #2114

markblundeberg committed:
Remove NULLDUMMY

Now that the Schnorr multisig upgrade is active, the NULLDUMMY standard
flag is totally superseded and pointless.

Test Plan:

make check
qa/pull-tester/rpc-tests.py
run IBD

This is a port of https://reviews.bitcoinabc.org/D4577

markblundeberg committed:
Make more script validation flags backward compatible

This is a backport of PR10699 ( https://github.com/bitcoin/bitcoin/pull/10699/files ),
but has to be adapted slightly since we don't just do softforking flags.

Pull request description:

  This change makes `SCRIPT_VERIFY_UPGRADABLE_NOPS` not apply to `OP_CHECKLOCKTIMEVERIFY` and `OP_CHECKSEQUENCEVERIFY`.
  This is a no-op as `UPGRADABLE_NOPS` is only set for mempool transactions, and those always have `SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY`
  and `SCRIPT_VERIFY_CHECKSEQUENCEVERIFY` set as well. The advantage is that setting more flags now always results
  in a reduction in acceptable scripts (=softfork).

  This results in a nice and testable property for validation, for which a new test is added.

  This also means that the introduction of a new definition for a NOP or witness version will likely need the following procedure (example OP_NOP8 here)
  * Remove OP_NOP8 from being affected by `SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS`.
  * Add a `SCRIPT_VERIFY_DISCOURAGE_NOP8`, which only applies to `OP_NOP8`.
  * Add a `SCRIPT_VERIFY_NOP8` which implements the new consensus logic.
  * Before activation, add `SCRIPT_VERIFY_DISCOURAGE_NOP8` to the mempool flags.
  * After activation, add `SCRIPT_VERIFY_NOP8` to both the mempool and consensus flags.

Test Plan:

make check
qa/pull-tester/rpc-tests.py

This is a (partial) port of https://reviews.bitcoinabc.org/D4703

sickpig committed:
Sync cashlib python test data structure with C++ ones

markblundeberg committed:
drop 'check3' upgrade-conditional-script-failure for Schnorr multisig

Now that upgrade has activatedi (Nov 20190), no need for forgiveness.

Test Plan

make check
qa/pull-tester/rpc-tests.py

This is a port o fhttps://reviews.bitcoinabc.org/D4576

sickpig committed:
Reformat MANDATORY_SCRIPT_VERIFY_FLAGS and add clang-format exception

markblundeberg committed:
add SCRIPT_VERIFY_MINIMALDATA to mandatory flags

Test Plan:

make check
qa/pull-tester/rpc-tests.py

This is a port of https://reviews.bitcoinabc.org/D4679

markblundeberg committed:
add SCHNORR_MULTISIG to mandatory flags

Test Plan:

make check
qa/pull-tester/rpc-tests.py

This is a port of https://reviews.bitcoinabc.org/D4673

sickpig commented:
Taking out of WIP

ptschip commented on qa/rpc-tests/minimaldata.py:
```diff
@@ -100,10 +101,10 @@ def build_block(self, parent, transactions=(), nTime=None):
-    def check_for_no_ban_on_rejected_tx(self, tx, reject_reason):
+    def check_for_ban_on_rejected_tx(self, tx, reject_reason=None):
         """Check we don't trigger a ban when sending a txn that the node rejects."""
         self.p2p.send_txs_and_test(
```
-Check we don't trigger
+Check we trigger

sickpig replied:
nice catch will fix

sickpig replied:
done

ptschip commented on src/txadmission.cpp:
```diff
@@ -703,11 +703,6 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
-    if (IsNov2019Activated(chainparams.GetConsensus(), chainActive.Tip()))
-    {
-        featureFlags |= SCRIPT_ENABLE_SCHNORR_MULTISIG;
-        featureFlags |= SCRIPT_VERIFY_MINIMALDATA;
```
I don't see where SCRIPT_ENABLE_SCHNORR_MULTISIG is added to any of either the Standard or Manadatory flags (in policy.h or script/standard.h), SCRIPT_VERIFY_MINIMALDATA however is defined in the Standard flags...also, in the lines above you could probably take out "featureFlags |= SCRIPT_ENABLE_CHECKDATASIG;"  since it's already defined in Standard flags?

sickpig replied:
both are added to the set of mandatory script in `standard.h`

see https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/2104/files#diff-117f0ad2885463dcf2b4cf1a5038939eR45

I should have removed `SCRIPT_VERIFY_MINIMALDATA` from policy.h 

WRT `SCRIPT_ENABLE_CHECKDATASIG` you are probably correct but I have to look more carefully.

ptschip replied:
You're right, I missed that..

sickpig replied:
you were right, I removed the `featureFlags |= SCRIPT_ENABLE_CHECKDATASIG` just left a stub for future features flags addition

sickpig committed:
Remove SCRIPT_VERIFY_MINIMALDATA from the set of standard flags

since the flag above as been added to the set of mandatory flags
and the set of standard flags is a strict superset of the mandatory
ones.

sickpig committed:
Fix a comment minimaldata.py python test

sickpig committed:
No need to explicitly add SCRIPT_ENABLE_CHECKDATASIG

The above flag has been added to the set of standard validation flag
after Nov 2019 network upgraded has been validate and check-pointed.

Just left a stub there for as guide for future features flags addition

ptschip commented on src/txadmission.cpp:
```diff
@@ -699,9 +699,10 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
     uint32_t featureFlags = 0;
-    if (IsNov2018Activated(chainparams.GetConsensus(), chainActive.Tip()))
+    if (IsMay2020Enabled(chainparams.GetConsensus(), chainActive.Tip()))
     {
```
If you're going to leave a stub you should maybe comment it out entirely and then add a comment so the next dev knows doesn't forget that this isn't just useless code to be removed.

sickpig committed:
[travis] remove BU special casing to avoid exiting before the timeout

Sometimes it seems that even if you have a very long time
before the timeout kicks in, it is better to exit earlier
while saving the cache rather than restarting from scratch
in case of failures
